File: backend/Models/JobApplication.cs
using System.ComponentModel.DataAnnotations;

namespace backend.Models;

public class JobApplication
{
    public int Id { get; set; }

    public int UserId { get; set; }

    public ApplicationUser? User { get; set; }

    [Required]
    [MaxLength(100)]
    public string CompanyName { get; set; } = string.Empty;

    [Required]
    [MaxLength(150)]
    public string JobTitle { get; set; } = string.Empty;

    [MaxLength(150)]
    public string? Location { get; set; }

    [MaxLength(500)]
    public string? JobLink { get; set; }

    public JobApplicationStatus Status { get; set; }
        = JobApplicationStatus.Saved;

    public DateOnly? AppliedDate { get; set; }

    public DateOnly? NextFollowUpDate { get; set; }

    [MaxLength(2000)]
    public string? Notes { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

File: frontend/src/types/jobApplication.ts
export const JOB_APPLICATION_STATUSES = [
	"Saved",
	"Applied",
	"OnlineAssessment",
	"Interview",
	"Offer",
	"Rejected",
	"Withdrawn",
] as const;

export type JobApplicationStatus = (typeof JOB_APPLICATION_STATUSES)[number];

export interface JobApplicationResponse {
	id: number;
	companyName: string;
	jobTitle: string;
	location: string | null;
	jobLink: string | null;
	status: JobApplicationStatus;
	appliedDate: string | null;
	nextFollowUpDate: string | null;
	notes: string | null;
	createdAt: string;
	updatedAt: string;
}

export interface JobApplicationUpsertRequest {
	companyName: string;
	jobTitle: string;
	location: string | null;
	jobLink: string | null;
	status: JobApplicationStatus;
	appliedDate: string | null;
	nextFollowUpDate: string | null;
	notes: string | null;
}

export interface JobApplicationDraft {
	companyName: string;
	jobTitle: string;
	location: string;
	jobLink: string;
	status: JobApplicationStatus;
	appliedDate: string;
	nextFollowUpDate: string;
	notes: string;
}

export type JobApplicationFieldName = keyof JobApplicationDraft;
export type JobApplicationFieldErrors = Partial<Record<JobApplicationFieldName, string>>;

export const EMPTY_JOB_APPLICATION_DRAFT: JobApplicationDraft = {
	companyName: "",
	jobTitle: "",
	location: "",
	jobLink: "",
	status: "Saved",
	appliedDate: "",
	nextFollowUpDate: "",
	notes: "",
};

export function toJobApplicationUpsertRequest(
	draft: JobApplicationDraft,
): JobApplicationUpsertRequest {
	return {
		companyName: draft.companyName.trim(),
		jobTitle: draft.jobTitle.trim(),
		location: normalizeOptionalText(draft.location),
		jobLink: normalizeOptionalText(draft.jobLink),
		status: draft.status,
		appliedDate: normalizeOptionalText(draft.appliedDate),
		nextFollowUpDate: normalizeOptionalText(draft.nextFollowUpDate),
		notes: normalizeOptionalText(draft.notes),
	};
}

export function toJobApplicationDraft(
	application: JobApplicationResponse,
): JobApplicationDraft {
	return {
		companyName: application.companyName,
		jobTitle: application.jobTitle,
		location: application.location ?? "",
		jobLink: application.jobLink ?? "",
		status: application.status,
		appliedDate: application.appliedDate ?? "",
		nextFollowUpDate: application.nextFollowUpDate ?? "",
		notes: application.notes ?? "",
	};
}

function normalizeOptionalText(value: string): string | null {
	const normalizedValue = value.trim();

	return normalizedValue ? normalizedValue : null;
}
